packages/backend/src/models/DashboardModel/DashboardModel.ts

```diff
@@ -876,4 +879,180 @@ export class DashboardModel {
+                DashboardsTableName,
+                `${DashboardsTableName}.dashboard_id`,
+                '=',
+                `${DashboardVersionsTableName}.dashboard_id`,
+            )
+            .where(`${DashboardsTableName}.dashboard_uuid`, dashboardUuid)
+            .andWhere(
+                `${DashboardTilesTableName}.dashboard_tile_uuid`,
+                dashboardTileUuid,
+            )
+            .select(`${DashboardTilesTableName}.dashboard_tile_uuid`)
+            .first();
+
+        if (!dashboardTile) {
+            throw new NotFoundError('Dashboard tile not found for dashboard');
+        }
+    }
+
+    async createComment(
+        dashboardUuid: string,
+        dashboardTileUuid: string,
+        text: string,
+        replyTo: string | null,
+        user: LightdashUser,
+    ): Promise<string> {
+        await this.checkDashboardTileExistsInDashboard(
+            dashboardUuid,
+            dashboardTileUuid,
+        );
+        const [{ comment_id: commentId }] = await this.database(
+            DashboardTileCommentsTableName,
+        )
+            .insert({
+                text,
+                dashboard_tile_uuid: dashboardTileUuid,
+                reply_to: replyTo ?? null,
+                user_uuid: user.userUuid,
+            })
+            .returning('comment_id');
+
+        return commentId;
+    }
+
+    async findCommentsForDashboard(
+        dashboardUuid: string,
+        userUuid: string,
+        canUserRemoveAnyComment: boolean,
+    ): Promise<Record<string, Comment[]>> {
+        const dashboard = await this.getById(dashboardUuid);
+
+        const tileUuids = dashboard.tiles.map((tile) => tile.uuid);
+
+        const commentsWithUsers = await this.database(
+            DashboardTileCommentsTableName,
+        )
+            .leftJoin(
+                UserTableName,
+                `${DashboardTileCommentsTableName}.user_uuid`,
+                '=',
+                `${UserTableName}.user_uuid`,
+            )
+            .select(
+                `${DashboardTileCommentsTableName}.*`,
+                `${UserTableName}.first_name`,
+                `${UserTableName}.last_name`,
+                `${DashboardTileCommentsTableName}.dashboard_tile_uuid`,
+            )
+            .whereIn(
+                `${DashboardTileCommentsTableName}.dashboard_tile_uuid`,
+                tileUuids,
+            )
+            .andWhere(`${DashboardTileCommentsTableName}.resolved`, false)
+            .orderBy(`${DashboardTileCommentsTableName}.created_at`, 'asc');
+
+        const commentsPerDashboardTile: Record<string, Comment[]> = {}; // Stores comments grouped by their dashboard_tile_uuid
+        const allComments: Record<string, Comment> = {}; // Fast access lookup for parent comments
+        const orphanReplies: Record<string, Comment[]> = {}; // Stores orphan replies keyed by their intended parent's commentId
+
+        commentsWithUsers.forEach((comment) => {
```

Comment:
this is now in only one loop - I've left comments to ensure it's clear, but it might not be. LMK what you think

Reply:
that's better, thanks. Could we extract this method outside of `findCommentsForDashboard` ? I need to also parse some results in another db query and we could reuse this. We can call it `parseComments`

Reply:
created `parseComments` method here: https://github.com/lightdash/lightdash/pull/9018/commits/f3b8760257be3f80ffd22ef737fee68c59cd72d7 LMK What you think!